### PR TITLE
Add initial /admin endpoint.

### PR DESCRIPTION
**Pull request by fredrikekre**
This patch adds an initial /admin endpoint. To alleviate the need for authentication this endpoint only accepts requests from localhost. Currently the only functionality of the endpoint is to toggle debug logging for the running server process.

### CODE REVIEW

File: src/admin.jl [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+const admin_lock = ReentrantLock()
+original_logger = nothing
+
+# Simple admin tasks
+function handle_admin(http::HTTP.Stream)
+    return @lock admin_lock handle_admin_locked(http)
+end
+function handle_admin_locked(http::HTTP.Stream)
+    # Forbid all /admin requests not from localhost
+    if http.stream.c.peerip != Sockets.localhost
```
Comment (staticfloat) [inline src/admin.jl]: We need to test that this properly respects headers such as `X-Real-Ip` and whatnot, as if the PkgServer is placed behind an nginx TLS-terminating reverse proxy, the socket will _always_ look like it's connected to localhost.  ;)

Of course if we ever have something more important than just toggling logging put here, we'll definitely need to implement some kind of cryptographic authentication.  HTTP digest authentication should be sufficient, as we're using TLS on all PkgServers.

File: src/admin.jl [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        HTTP.setstatus(http, 403)
+        msg = "The /admin endpoint is only allowed from localhost ($(Sockets.localhost)).\n"
+        HTTP.setheader(http, "Content-Length" => string(sizeof(msg)))
+        HTTP.startwrite(http)
+        HTTP.write(http, msg)
+        return
+    end
+    # TODO: Currently everything hitting /admin just toggles debug output.
```
Comment (staticfloat) [inline src/admin.jl]: I'd prefer to build for the future and start off by implementing `/admin/logging` (where a `GET` request returns the current logging status, and a `POST` request sets the current logging status).  In general, we should avoid toggles and instead use idempotent actions like a `POST` where we set `logging=true`.

I think `/admin/shutdown` would also be useful, for remotely rebooting PkgServers (but that should probably only be implemented after more secure cryptography)

File: src/admin.jl (hunk 1 above)
Comment (fredrikekre) [reply src/admin.jl]: True, I only considered the current docker setup, where nginx requests come from the docker interface and not localhost.

**Comment by DilumAluthge** [comment]
This all seems like a pretty big security footgun, right? Why does any of this need to be exposed in the server itself?

**Comment by staticfloat** [comment]
> This all seems like a pretty big security footgun, right?

In what way is it a security footgun?  Although I agree that we should not store the token in an `ENV` and should instead store it in a file to be read upon first use, I don't see how setting/getting logging level is a security issue even if the token does get leaked.

> Why does any of this need to be exposed in the server itself?

How do you suggest altering the behavior of a running server?  I think sending HTTP requests to alter the behavior is pretty standard.

**Comment by fredrikekre** [comment]
I updated the PR to use `htpasswd` instead. (In docker) you would only mount a `.htpasswd` file with `bcrypt` hashed password. Still have to flesh out what endpoints should be available under `/admin`. I think `/admin/logging/tail` would be kinda neat to basically just `tail -f` the logs back, but HTTP.jl is very spammy with log messages so need to filter out them in that case.

**Comment by fredrikekre** [comment]
From my POV this is good to go now.

As mentioned above, this now uses [htpasswd](https://httpd.apache.org/docs/current/programs/htpasswd.html) for verifying the authenticating user. You only need to supply a password file which will include a bcrypt hashed password to compare incoming requests against.

The `GET /admin/logs` endpoint attaches to a attachable logger and stream log messages back to the client. The `level` query parameter can be used to enable debug logs.

All of this is disabled by default and I have added warnings to not enable this if not using HTTPs.

**Comment by staticfloat** [comment]
I think it makes sense to drop the `htpasswd` dependency and just store a `sha256()`-hashed version of the password, then do the hashing in pure Julia.  It should be faster and continue to keep our dependencies minimal.  Also, it probably makes sense to upstream the attaching logger to `LoggingExtras` eventually, but let's use it for a bit and shake out any bugs/extra features it needs first.

**Pull request merged**